src/shared-validator.h

```diff
@@ -180,11 +181,14 @@ class SharedValidator {
  private:
   struct FuncType {
     FuncType() = default;
-    FuncType(const TypeVector& params, const TypeVector& results)
-        : params(params), results(results) {}
+    FuncType(const TypeVector& params,
+             const TypeVector& results,
+             Index type_index)
+        : params(params), results(results), type_index(type_index) {}
 
     TypeVector params;
     TypeVector results;
+    Index type_index;
```

Comment:
Why does `FuncType` need to store its own index now if it didn't need to before?

Reply:
previously we don't care about the function type indices because care only about function signatures.
You can get function signature via function index:
https://github.com/WebAssembly/wabt/blob/9aac0a28896a714b698ab489a473b9bc05deb7c3/src/shared-validator.h#L293

But there is no way to get function type index via function index and so, I preserve this in FuncType.